obj: async post-commit

pbalcer opened the pull request:
Ref: pmem/issues#381

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1671)
<!-- Reviewable:end -->

krzycz commented:
Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/libpmemobj/obj.c, line 1027 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-KdaB4JgsACt1JRw2ZUZ:-KdaB4JgsACt1JRw2ZU_:b2ngwtm) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/obj.c#L1027)):*
> ```C
> 		obj_pool_init();
> 
> 		if ((pop->tx_postcommit_tasks = ringbuf_new(0)) == NULL) {
> ```

0 ?

---

*[src/libpmemobj/ringbuf.c, line 78 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda0vGgmILTNox2IwnB:-Kda0vGhi48TGJTdcGWS:b-efikrk) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/ringbuf.c#L78)):*
> ```C
> 		return NULL;
> 
> 	sem_init(&rbuf->nfree, 0, length);
> ```

missing error handling (same applies to all the calls to sem_*() APIs - sem_wait/sem_post/...)

---

*[src/libpmemobj/ringbuf.h, line 34 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda3qkA9sxBQoh2f2qE:-Kda3qkA9sxBQoh2f2qF:b-eteot7) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/ringbuf.h#L34)):*
> ```objc
> 
> /*
>  * ringbuf.h -- internal definitions for mpsc ring buffer
> ```

mp_M_c?

---

*[src/libpmemobj/tx.c, line 806 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda4IKHPFXqmNWRzQry:-Kda4IKHPFXqmNWRzQrz:b-swks9h) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/tx.c#L806)):*
> ```C
> tx_get_pop(void)
> {
> 	return tx.pop;
> ```

Hmmm, don't we need to check if stage is NONE?

---

*[src/libpmemobj/tx.c, line 1279 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda6AF3JmDg3T0ncMHG:-Kda6AF4N8JQofLp4BQT:b-pa5es6) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/tx.c#L1279)):*
> ```C
> 
> /*
>  * tx_postcommit_cleanup -- performs all the necessery cleanup on a lane after
> ```

necessary

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Debug traces might be helpful sometimes...

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-KdaGV4nio2QRbwFN-hH:b-ej7ics)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
I'll update this patch once CTL is merged. I've provided some preliminary performance results here:
pmem//issues#381

plebioda commented:
Can't review this patch because it is not completed.

---

Reviewed 13 of 14 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/libpmemobj/obj.c, line 1027 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-KdaB4JgsACt1JRw2ZUZ:-KfGYZiU-Tg3W5eJQOoA:b-sik73a) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/obj.c#L1027)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

0 ?
</blockquote></details>

bump ?

---

*[src/libpmemobj/ringbuf.h, line 46 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-KfGX56l_73A4Y0Y5z7T:-KfGX56l_73A4Y0Y5z7U:b-ktp1nq) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/ringbuf.h#L46)):*
> ```objc
> struct ringbuf;
> 
> struct ringbuf *ringbuf_new(unsigned length);
> ```

size_t ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-KfGZgmZDHoz_VJuu5pl:by6ynid)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
It is complete to the point to which it it possible.

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-Kfek7o5SPcyXw5U1_f7:b-51026g)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
I have updated this patch with the CTL entry point and related changes. I'll separate out lane-related optimizations to a different patch and fix the windows build soon.

krzycz commented:
Reviewed 13 of 16 files at r2.
Review status: 16 of 19 files reviewed at latest revision, 6 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->

krzycz commented:
1) It would be nice to have some example and/or benchmark that illustrates how to use this feature and/or measures the performance gain.
My first thought was to simply re-use the existing examples/benchmarks, but run it with some non-default CTL settings (env var), but I realized it's not that easy in case of async post-commit, as you need to explicitly run the PC thread.

2) Please, use LOG/ERR macros!!!

---

Reviewed 4 of 10 files at r3, 42 of 42 files at r4.
Review status: all files reviewed at latest revision, 37 unresolved discussions, some commit checks failed.

---

*[doc/libpmemobj.3.md, line 2262 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCPzGoAlNwDqfVmr92:-KjCPzGoAlNwDqfVmr93:b67q0lm) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2262)):*
> ```Markdown
> tx.post_commit.queue_depth | rw | - | int | int | integer
> 
> Controls the depth of the post-commit tasks queue. A post-commit task is the
> ```

What if I configure queue_depth to some value, but no PC thread is running?
Would it fall-back to the normal mode, like w/o any post-commit queue?

---

*[doc/libpmemobj.3.md, line 2276 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC_KLqSUqFblQhKiRm:-KjC_KLrJRs5VWRUfKFL:b-741g9m) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2276)):*
> ```Markdown
> will happen if the queue depth value is too large.
> 
> As a general rule, this value should be set to around 1024 - average number of
> ```

...and it must be a power of two

---

*[doc/libpmemobj.3.md, line 2283 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCT_v-4gW2lMBruY42:-KjCT_v-4gW2lMBruY43:b-4hnv2g) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2283)):*
> ```Markdown
> currently executed.
> 
> Always returns 0.
> ```

unless the ringbuf allocation fails...
or if the value is not a power of two...

---

*[doc/libpmemobj.3.md, line 2285 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCOVg-vEv3_BON7ZGc:-KjCOVg-vEv3_BON7ZGd:b-kj0hme) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2285)):*
> ```Markdown
> Always returns 0.
> 
> tx.post_commit.worker | r- | - | void * | - | -
> ```

It appears a bit counter-intuitive to me that one need to issue read (get) to trigger the working function. For me, a write (set) would be better (value could be ignored). Same applies to `tx.post_commit.stop`.
Or, if it's `get`, then let it return something (i.e. a number of committed transactions, etc.)

---

*[doc/libpmemobj.3.md, line 2294 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCOUKFCcOXj9yP6lJa:-KjCOUKFCcOXj9yP6lJb:b-ecnaza) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2294)):*
> ```Markdown
> Always returns 0.
> 
> tx.post_commit.worker | r- | - | void * | - | -
> ```

.stop ?

---

*[doc/libpmemobj.3.md, line 2296 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCQI4iY42h47NyiH4A:-KjCQI4iY42h47NyiH4B:brrsm93) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2296)):*
> ```Markdown
> tx.post_commit.worker | r- | - | void * | - | -
> 
> This function forces all the post-commit worker functions to exit and return
> ```

What happens with the tasks in the post-commit queue? Does it have to be empty?
Shall we handle it somehow?

---

*[src/libpmemobj/lane.c, line 446 at r2](https://reviewable.io:443/reviews/pmem/nvml/1671#-KiGzx5iRbDs8umt5K0y:-KiGzx5iRbDs8umt5K0z:b-6oleat) ([raw file](https://github.com/pmem/nvml/blob/09ae4ceb6577ab629c249dfbd2c71eaf8946c2c0/src/libpmemobj/lane.c#L446)):*
> ```C
> }
> 
> void
> ```

missing brief

---

*[src/libpmemobj/ringbuf.c, line 71 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCVMlpquhN8ymc4Nxx:-KjCVMlpquhN8ymc4Nxy:b6t6vwr) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L71)):*
> ```C
> {
> 	/* length must be a power of two due to masking */
> 	ASSERT(util_popcount(length) <= 1);
> ```

Was util_popcount() added only to be used in this assertion?
What about: `ASSERTeq(length & (~length + 1), length); /* is power of 2 */`

---

*[src/libpmemobj/ringbuf.c, line 122 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCWFcN5Sgfwyg27BKv:-KjCWFcOw30O-sxSEfgf:b-je1zdm) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L122)):*
> ```C
> 
> 	/* XXX just unlock all waiting threads somehow... */
> 	for (uint64_t i = 0; i < 1024; ++i)
> ```

1024? Don't we have a named constant for that.  Or better read the number of lanes from the pool descriptor.

---

*[src/libpmemobj/ringbuf.c, line 134 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCXFHujfBSDBxDDgPO:-KjCXFHujfBSDBxDDgPP:buuboub) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L134)):*
> ```C
> 	os_semaphore_delete(rbuf->nfree);
> 	os_semaphore_delete(rbuf->nused);
> 	Free(rbuf);
> ```

What if there are any tasks enqueued? An assertion is a minimum:
`ASSERTeq(read-pos, write_pos);` ?

---

*[src/libpmemobj/ringbuf.c, line 162 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCWwLNTtIAdEMIYnIc:-KjCWwLOKZh_UVRyOkn4:b-q3qasb) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L162)):*
> ```C
>  */
> int
> ringbuf_enqueue(struct ringbuf *rbuf, void *data)
> ```

Is it actually used anywhere (except the unit tests)? Isn't "tryenqueue" enough?

---

*[src/libpmemobj/ringbuf.c, line 222 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCeSdRqQjKnWMZeZ6P:-KjCXeC4eaFuU3V86MMe:b-5np7md) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L222)):*
> ```C
>  */
> void *
> ringbuf_dequeue(struct ringbuf *rbuf)
> ```

not used in other files (could be static / merged with dequeue_s) 

---

*[src/libpmemobj/ringbuf.c, line 239 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCd7e3iOw8RWnubfL4:-KjCd7e3iOw8RWnubfL5:b-r6iy2m) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L239)):*
> ```C
>  */
> void *
> ringbuf_trydequeue(struct ringbuf *rbuf)
> ```

not used (except for unit test)

---

*[src/libpmemobj/ringbuf.c, line 255 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCdjCvhBMJChgqmo_p:-KjCdjCvhBMJChgqmo_q:b-9q54b5) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L255)):*
> ```C
>  */
> void *
> ringbuf_trydequeue_s(struct ringbuf *rbuf, size_t data_size)
> ```

not used

---

*[src/libpmemobj/ringbuf.c, line 268 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCd-u69L6ftgpqK1Cs:-KjCd-u7mfggJIa_46MO:bo601tl) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L268)):*
> ```C
>  */
> void *
> ringbuf_dequeue_s(struct ringbuf *rbuf, size_t data_size)
> ```

As I can see, only the "_s" variant is actually used in tx.c. So, what's the point in having "ringbuf_dequeue"?

---

*[src/libpmemobj/ringbuf.c, line 271 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCY3iQG7IXmx5-hd9a:-KjCY3iQG7IXmx5-hd9b:bmoxoew) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L271)):*
> ```C
> {
> 	void *r = ringbuf_dequeue(rbuf);
> 	VALGRIND_ANNOTATE_NEW_MEMORY(r, data_size);
> ```

Could you add more info/comment on why this is needed?

---

*[src/libpmemobj/tx.c, line 2211 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCUAPUIW2ehXII-I8d:-KjCUAPUIW2ehXII-I8e:bd1c885) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/tx.c#L2211)):*
> ```C
> 	int arg_in = *(int *)arg;
> 
> 	struct ringbuf *ntasks = ringbuf_new((unsigned)arg_in);
> ```

Any check if arg_in value is a power of two?
There is an assertion in ringbuf_new(), but no other handling of invalid length.

---

*[src/libpmemobj/tx.c, line 2216 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCT5iWCesxkRwPpUl_:-KjCT5iWCesxkRwPpUla:b-eiqj1l) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/tx.c#L2216)):*
> ```C
> 
> 	if (pop->tx_postcommit_tasks != NULL) {
> 		ringbuf_delete(pop->tx_postcommit_tasks);
> ```

What if it wasn't empty?

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 34 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCCiDej1izCIo21SaZ:-KjCCiDej1izCIo21Sa_:b-pl7aji) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L34)):*
> ```C
> 
> /*
>  * obj_async_postcommit.c -- tests for the ctl entry points: prefault
> ```

Pls, fix the description.
Also, some more detailed description of the test scenario is welcome.
As I understand, you first spawn 4 PC threads, waiting for TX to be committed. Then another 4 worker threads, each performing 1000 transactions (4000 in total).
PC threads are expected to do post-commit cleanups in parallel to the worker threads, but how this test verifies it really takes place? How does it verify the non-zero queue depth has any effect?

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 87 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCD-0yRGph7allutpu:-KjCD-0zk_4vkpv57RZO:b-96lew7) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L87)):*
> ```C
> 
> 	if (argc != 2)
> 		UT_FATAL("usage: %s file-name",
> ```

no need to break the line

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 95 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCD4zrhbYrEhpYjGh4:-KjCD4zrhbYrEhpYjGh5:b-6q4aa5) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L95)):*
> ```C
> 	if ((pop = pmemobj_create(path, LAYOUT, PMEMOBJ_MIN_POOL * 10,
> 			S_IWUSR | S_IRUSR)) == NULL)
> 			UT_FATAL("!pmemobj_create: %s", path);
> ```

unindent by 1 tab

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 99 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCDIVJ493C742jUU3T:-KjCDIVJ493C742jUU3U:b-p5tfqo) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L99)):*
> ```C
> 	pthread_t pc_worker[PC_WORKERS];
> 
> 	if (PC_WORKERS) {
> ```

if (true) ?

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 103 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCDX_dDdkM3YFTeL7s:-KjCDX_dDdkM3YFTeL7t:b-e0tol5) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L103)):*
> ```C
> 		pmemobj_ctl_set(pop, "tx.post_commit.queue_depth", &qdepth);
> 		for (int i = 0; i < PC_WORKERS; ++i) {
> 			pthread_create(&pc_worker[i],
> ```

PTHREAD_CREATE()

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 128 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCEtbt2ELZSYs7KhBZ:-KjCEtbt2ELZSYs7KhB_:b-7ropln) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L128)):*
> ```C
> 	}
> 
> 	if (PC_WORKERS) {
> ```

ditto

---

*[src/test/obj_async_postcommit/obj_async_postcommit.filters, line 1 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCCPyBRWyLvQZuO_zC:-KjCCPyBRWyLvQZuO_zD:bkuua2e) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.filters#L1)):*
> ```XML
> <?xml version="1.0" encoding="utf-8"?>
> ```

*.vcxproj.filters"

---

*[src/test/obj_async_postcommit/obj_async_postcommit.vcxproj, line 69 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCMru7ylAx7ee0jiAm:-KjCMru7ylAx7ee0jiAn:b-531ftn) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.vcxproj#L69)):*
> ```XML
>     <ClCompile>
>       <PrecompiledHeader>NotUsing</PrecompiledHeader>
>       <CompileAs>CompileAsCpp</CompileAs>
> ```

Is that really needed (compiling as C++)?

---

*[src/test/obj_async_postcommit/TEST0.PS1, line 41 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCC0785uR3J6HVcmJN:-KjCC0785uR3J6HVcmJO:b-u1i2aj) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/TEST0.PS1#L41)):*
> ```
>     $DIR = ""
>     )
> $ENV:UNITTEST_NAME = "obj_pvector\TEST0"
> ```

obj_pvector ?

---

*[src/test/obj_ringbuf/obj_ringbuf.filters, line 1 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCAeJoVP8YxQNZEFdQ:-KjCAeJoVP8YxQNZEFdR:bspzmm5) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.filters#L1)):*
> ```XML
> <?xml version="1.0" encoding="utf-8"?>
> ```

"obj_ringbuf.vcxproj.filters"

---

*[src/test/obj_ringbuf/obj_ringbuf.vcxproj, line 50 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCAHe_6MXyXQvj04yT:-KjCAHe_6MXyXQvj04yU:bo3tdzf) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.vcxproj#L50)):*
> ```XML
>   <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
>     <ClCompile>
>       <Optimization>Disabled</Optimization>
> ```

optimization settings should be taken from test_xxx.props

---

*[src/test/obj_ringbuf/obj_ringbuf.vcxproj, line 93 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC9HCE2pHnHsS8vYGe:-KjC9HCGf0D-xzGtYRXr:bz05h7r) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.vcxproj#L93)):*
> ```XML
>       <Project>{492baa3d-0d5d-478e-9765-500463ae69aa}</Project>
>     </ProjectReference>
>     <ProjectReference Include="..\..\libpmemobj\libpmemobj.vcxproj">
> ```

Does it still depend on libpmemobj, if most (if not all) of the obj files are compiled/linked directly?

---

*[src/test/obj_ringbuf/TEST0, line 44 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC98MGxA9xq8s3KyyU:-KjC98MGxA9xq8s3KyyV:b-38uw1z) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/TEST0#L44)):*
> ```
> 
> require_test_type medium
> ```

require_fs_type any

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-KjC_cTZNc2i7Z7iLUXx:b-b49etp)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
1. The benefit from this is very workload dependent, and so the benchmarks we have won't show much improvements. I'll create a simple example as a separate pull request. The test shows you how to use it.

2.  I don't see much benefit in using LOG macros at the entrance of every function, we are effectively creating a poor man'd gdb, but fine, I'll try to remember about this ;)
Error strings and errno should be only set at the outermost layer of the library, otherwise we will end up with a mess of error indicators being overwritten all over the place.

---

Review status: 40 of 53 files reviewed at latest revision, 37 unresolved discussions.

---

*[doc/libpmemobj.3.md, line 2262 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCPzGoAlNwDqfVmr92:-KjINlIaSVbm166-DQcr:b-1o785h) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2262)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

What if I configure queue_depth to some value, but no PC thread is running?
Would it fall-back to the normal mode, like w/o any post-commit queue?
</blockquote></details>

Why would it? It does what you would expect: there's no special behavior associated. And how would you detect it anyway?

Maybe I'm not understanding your question?

If you don't launch the worker entry point, you will never cleanup the lanes data and thus will end up several lanes locked up until you do launch the threads. I think that is the least surprising thing to do.

---

*[doc/libpmemobj.3.md, line 2276 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC_KLqSUqFblQhKiRm:-KjIKGRK1piTmjaZRGwM:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2276)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

...and it must be a power of two
</blockquote></details>

Done.

---

*[doc/libpmemobj.3.md, line 2283 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCT_v-4gW2lMBruY42:-KjIKH9VZkJNx9XpKbhc:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2283)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

unless the ringbuf allocation fails...
or if the value is not a power of two...
</blockquote></details>

Done.

---

*[doc/libpmemobj.3.md, line 2285 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCOVg-vEv3_BON7ZGc:-KjIKHputPi2EAdMHnpB:bch47x6) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2285)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

It appears a bit counter-intuitive to me that one need to issue read (get) to trigger the working function. For me, a write (set) would be better (value could be ignored). Same applies to `tx.post_commit.stop`.
Or, if it's `get`, then let it return something (i.e. a number of committed transactions, etc.)
</blockquote></details>

write functions can be set from config and require a parser for the argument. In the very first version of the CTL patch (and in the initial design) there was a specific "trigger" type, but it got axed when everyone suggested to use separate `ctl_get`/`ctl_set`.

---

*[doc/libpmemobj.3.md, line 2294 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCOUKFCcOXj9yP6lJa:-KjIKmpE4vQ13Az2bVov:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2294)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

.stop ?
</blockquote></details>

Done.

---

*[doc/libpmemobj.3.md, line 2296 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCQI4iY42h47NyiH4A:-KjIKnD7x55mg8rHKt0R:byccnm1) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2296)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

What happens with the tasks in the post-commit queue? Does it have to be empty?
Shall we handle it somehow?
</blockquote></details>

It doesn't have to be empty, and nothing would have happened, the lanes would have been simply cleaned up on recovery.
But nevertheless, I've changed the implementation a bit so that the stop method now waits for all pending tasks to be completed.

---

*[src/libpmemobj/lane.c, line 446 at r2](https://reviewable.io:443/reviews/pmem/nvml/1671#-KiGzx5iRbDs8umt5K0y:-KjIL4Ghi_o816MsUcsf:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/09ae4ceb6577ab629c249dfbd2c71eaf8946c2c0/src/libpmemobj/lane.c#L446)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing brief
</blockquote></details>

Done.

---

*[src/libpmemobj/obj.c, line 1027 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-KdaB4JgsACt1JRw2ZUZ:-KjIL585RWZGqmopCLse:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/obj.c#L1027)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

bump ?
</blockquote></details>

Done.

---

*[src/libpmemobj/ringbuf.c, line 78 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda0vGgmILTNox2IwnB:-KjIL6Twqv9nzW-pPO2i:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/ringbuf.c#L78)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

missing error handling (same applies to all the calls to sem_*() APIs - sem_wait/sem_post/...)
</blockquote></details>

Done.

---

*[src/libpmemobj/ringbuf.c, line 71 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCVMlpquhN8ymc4Nxx:-KjIN3T3b0g3NbCriFMn:b5pmjmm) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L71)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Was util_popcount() added only to be used in this assertion?
What about: `ASSERTeq(length & (~length + 1), length); /* is power of 2 */`
</blockquote></details>

Yes. Does it hurt us to add this function?
I'd argue my way is more readable and faster.

---

*[src/libpmemobj/ringbuf.c, line 122 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCWFcN5Sgfwyg27BKv:-KjINIbN1ld1g5SCuAq0:b-fvl9be) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L122)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

1024? Don't we have a named constant for that.  Or better read the number of lanes from the pool descriptor.
</blockquote></details>

This is generic code, I don't want to make a dependency on lanes.
It's still not ideal, but I don't think there's a better way that doesn't involve adding extraneous variables or timed waits.

---

*[src/libpmemobj/ringbuf.c, line 134 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCXFHujfBSDBxDDgPO:-KjIL7gERm5QTFRAahF6:bwf3rr6) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L134)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

What if there are any tasks enqueued? An assertion is a minimum:
`ASSERTeq(read-pos, write_pos);` ?
</blockquote></details>

Initially I wanted to reply that this is implementation dependent, and the user might want to delete the ringbuf with its contents - the ring buffer itself doesn't associate any of its own resources with the user data and so there's nothing that might leak.
But then I decided that in 99% of the use cases, we would only want to call delete on an empty container, so I added it, we can always change it.

---

*[src/libpmemobj/ringbuf.c, line 162 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCWwLNTtIAdEMIYnIc:-KjILYyWetI7hqeb6rsJ:b-bkogcc) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L162)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is it actually used anywhere (except the unit tests)? Isn't "tryenqueue" enough?
</blockquote></details>

All variants are provided for consistency sake.

---

*[src/libpmemobj/ringbuf.c, line 271 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCY3iQG7IXmx5-hd9a:-KjILjdbZrEWqg7xQ5f3:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L271)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Could you add more info/comment on why this is needed?
</blockquote></details>

Done.

---

*[src/libpmemobj/ringbuf.h, line 34 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda3qkA9sxBQoh2f2qE:-KjILkF61MQ0B-Z3jAyl:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/ringbuf.h#L34)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

mp_M_c?
</blockquote></details>

Done.

---

*[src/libpmemobj/ringbuf.h, line 46 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-KfGX56l_73A4Y0Y5z7T:-KjILkfzSHimiwSnYWO7:bb76mu4) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/ringbuf.h#L46)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

size_t ?
</blockquote></details>

Why?

---

*[src/libpmemobj/tx.c, line 806 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda4IKHPFXqmNWRzQry:-KjILoUr94PgRX0uHiHl:b3gbw3z) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/tx.c#L806)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Hmmm, don't we need to check if stage is NONE?
</blockquote></details>

Nope, the pop variable will be NULL if there's no transaction.

---

*[src/libpmemobj/tx.c, line 1279 at r1](https://reviewable.io:443/reviews/pmem/nvml/1671#-Kda6AF3JmDg3T0ncMHG:-KjIM0ObhWCwl2ipSK6A:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/367ebdc2b6c6485956595af5712773fe99434c3a/src/libpmemobj/tx.c#L1279)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

necessary
</blockquote></details>

Done.

---

*[src/libpmemobj/tx.c, line 2211 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCUAPUIW2ehXII-I8d:-KjIM2kyeOLvN_O3xMCG:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/tx.c#L2211)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Any check if arg_in value is a power of two?
There is an assertion in ringbuf_new(), but no other handling of invalid length.
</blockquote></details>

Done.

---

*[src/libpmemobj/tx.c, line 2216 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCT5iWCesxkRwPpUl_:-KjIM3P1acvMUV1UErPC:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/tx.c#L2216)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

What if it wasn't empty?
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 34 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCCiDej1izCIo21SaZ:-KjIM4mFFZxV6Ya9iBPl:bfjajvm) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L34)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Pls, fix the description.
Also, some more detailed description of the test scenario is welcome.
As I understand, you first spawn 4 PC threads, waiting for TX to be committed. Then another 4 worker threads, each performing 1000 transactions (4000 in total).
PC threads are expected to do post-commit cleanups in parallel to the worker threads, but how this test verifies it really takes place? How does it verify the non-zero queue depth has any effect?


</blockquote></details>

I've reworked this test somehow to fail in a more significant way if something breaks. If the post commit workers don't work it will now hang and eventually time out.
Apart from intercepting somehow library internals, there's really no other way to test if the cleanups are actually deferred. Or maybe you have any ideas?

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 87 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCD-0yRGph7allutpu:-KjIMkifL98vQ8TNtfBT:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L87)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

no need to break the line
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 95 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCD4zrhbYrEhpYjGh4:-KjIMlE2gugitNgxZgEL:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L95)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

unindent by 1 tab
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 99 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCDIVJ493C742jUU3T:-KjIMleG-HekjC69wKUI:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L99)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

if (true) ?
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 103 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCDX_dDdkM3YFTeL7s:-KjIMlydpaotXoJV-N6U:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L103)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

PTHREAD_CREATE()
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.c, line 128 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCEtbt2ELZSYs7KhBZ:-KjIMmUH2cCrEUn5xnHR:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.c#L128)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.vcxproj, line 69 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCMru7ylAx7ee0jiAm:-KjIMnGPi3bM3TG9F6t1:brux1ir) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.vcxproj#L69)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is that really needed (compiling as C++)?
</blockquote></details>

Does it matter? I copy/pasted the test project.

---

*[src/test/obj_async_postcommit/TEST0.PS1, line 41 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCC0785uR3J6HVcmJN:-KjIMrYRi3ATuN44VTl1:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/TEST0.PS1#L41)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

obj_pvector ?
</blockquote></details>

Done.

---

*[src/test/obj_ringbuf/obj_ringbuf.vcxproj, line 50 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCAHe_6MXyXQvj04yT:-KjIMsCpeMGzal3uKpOn:b-xq6rw6) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.vcxproj#L50)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

optimization settings should be taken from test_xxx.props
</blockquote></details>

I copy pasted your changes, it's like that in every project file I checked.

---

*[src/test/obj_ringbuf/obj_ringbuf.vcxproj, line 93 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC9HCE2pHnHsS8vYGe:-KjIMxCeFqSus3cioNpq:bpf9f2g) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.vcxproj#L93)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Does it still depend on libpmemobj, if most (if not all) of the obj files are compiled/linked directly?
</blockquote></details>

Again, does it matter? It's like that everywhere.

---

*[src/test/obj_ringbuf/TEST0, line 44 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC98MGxA9xq8s3KyyU:-KjIN-R-rUhOACs0MQ66:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/TEST0#L44)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

require_fs_type any
</blockquote></details>

Done.

---

*[src/test/obj_async_postcommit/obj_async_postcommit.filters, line 1 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCCPyBRWyLvQZuO_zC:-KjIN0rxAFUCatrHK9mC:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.filters#L1)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

*.vcxproj.filters"
</blockquote></details>

Done.

---

*[src/test/obj_ringbuf/obj_ringbuf.filters, line 1 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCAeJoVP8YxQNZEFdQ:-KjIN1VL_vuJbRXZrTMz:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.filters#L1)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

"obj_ringbuf.vcxproj.filters"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-KjIJ-si2luBEa8XkEL-:bsfagmf)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 12 of 15 files at r5.
Review status: 50 of 53 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[src/libpmemobj/ringbuf.c, line 71 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCVMlpquhN8ymc4Nxx:-KjtOfzhqNHPVCTTQW35:b-bm2xr2) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L71)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Yes. Does it hurt us to add this function?
I'd argue my way is more readable and faster.
</blockquote></details>

No, it doesn't. But instead of just one line, we have a 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: 50 of 53 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[src/libpmemobj/ringbuf.c, line 71 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCVMlpquhN8ymc4Nxx:-KjtTf2nf14year8yV8Q:b-kcp4x3) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L71)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

No, it doesn't. But instead of just one line, we have a 
</blockquote></details>

...a single line plus a few more in another file. Just a detail...
It's like - I can do `x = a * a`, but why not to have a new function`sqr(a)` that does exactly the same using the CPU instruction SQR (if there is one), and then do `x = sqr(a)`?
Is it more readable? I'd argue. IMHO the most important is the comment that says everything in both cases.
Is it faster? Maybe... For debug we don't really care. But, now it's used in non-debug as well, so it (theoretically) can make a difference. However..., how often do you expect `ringbuf_new()`will be called - 1000x per second or... just once? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 3 of 15 files at r5.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[doc/libpmemobj.3.md, line 2262 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCPzGoAlNwDqfVmr92:-KkAJNhxCBUBn06bKDx7:b5lyfrr) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/doc/libpmemobj.3.md#L2262)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Why would it? It does what you would expect: there's no special behavior associated. And how would you detect it anyway?

Maybe I'm not understanding your question?

If you don't launch the worker entry point, you will never cleanup the lanes data and thus will end up several lanes locked up until you do launch the threads. I think that is the least surprising thing to do.
</blockquote></details>

Each PC worker thread must be explicitly spawned, so you know if there are any of them running. But as I understand, the problem could be with the order of operations - whether you change the queue depth first, then spawn the PC threads (so for a moment, the queue depth is non-zero, but no PC threads are running), or if you start PC threads first, then change the queue depth (may lead to race conditions).

OK, let's leave it as is.

---

*[src/libpmemobj/ringbuf.c, line 71 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCVMlpquhN8ymc4Nxx:-KkAPH6wRgIydGvIa5Wg:bg7lhs1) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/libpmemobj/ringbuf.c#L71)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

...a single line plus a few more in another file. Just a detail...
It's like - I can do `x = a * a`, but why not to have a new function`sqr(a)` that does exactly the same using the CPU instruction SQR (if there is one), and then do `x = sqr(a)`?
Is it more readable? I'd argue. IMHO the most important is the comment that says everything in both cases.
Is it faster? Maybe... For debug we don't really care. But, now it's used in non-debug as well, so it (theoretically) can make a difference. However..., how often do you expect `ringbuf_new()`will be called - 1000x per second or... just once? 

</blockquote></details>

Never mind - Maybe I'm picky on that just because when I grew up, there was no CPU having popcnt instruction. So, every programmer had to know how to do such a check using only basic arith operations, and I could never think about using popcnt for that purpose...

---

*[src/test/obj_async_postcommit/obj_async_postcommit.vcxproj, line 69 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCMru7ylAx7ee0jiAm:-KkAIDboJozszJwbUOrb:b-63kord) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_async_postcommit/obj_async_postcommit.vcxproj#L69)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Does it matter? I copy/pasted the test project.
</blockquote></details>

Not a big deal - just asking.

---

*[src/test/obj_ringbuf/obj_ringbuf.vcxproj, line 50 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjCAHe_6MXyXQvj04yT:-KkAIbxnrdYbjrvTyqdG:b9wdw1s) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.vcxproj#L50)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

I copy pasted your changes, it's like that in every project file I checked.
</blockquote></details>

So, it looks like I need to fix vcxproj files... ;-)

---

*[src/test/obj_ringbuf/obj_ringbuf.vcxproj, line 93 at r4](https://reviewable.io:443/reviews/pmem/nvml/1671#-KjC9HCE2pHnHsS8vYGe:-KkAJ6gNnFcC0aph_e10:b-gs68rx) ([raw file](https://github.com/pmem/nvml/blob/955f731d2a02629d481edf0799cfd54aac268456/src/test/obj_ringbuf/obj_ringbuf.vcxproj#L93)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Again, does it matter? It's like that everywhere.
</blockquote></details>

As most of the *obj* tests **do** link to libpmemobj...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->

krzycz commented:
2) Keep in mind that these macros are not (only) for us, but also for the users of NVML or the programs/applications based on NVML. Perhaps they don't even know how to use debugger. ;-) Sometimes, the only we can expect from them in case of a failure is to run their app on debug build with debug traces enabled and send the logs back to us.
Also, in case of tests run on Travis/AppVeyor (where you can't log in and debug the problem) debug traces could be the only way to identify the root cause of the problem. 
And finally, sometimes the issue is not the program crash - which is easy to debug. It could be an error that results in graceful exit. Then the log could help you to see where to put the break point.

In general, I think it's very useful to have tracing points - as much precise as they could be. They don't affect non-debug builds, and for debug builds they can be disabled, so again, not a huge performance impact.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-KkATkoVP65KRCFVU-Uu:bwmtveu)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671#-:-KkAXNXW79miVJ6mMdSs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 6 of 10 files at r6.
Review status: 51 of 55 files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 4 of 10 files at r6.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1671)*
<!-- Sent from Reviewable.io -->
